authz/providers/azure/rbac/rbac.go

```diff
@@ -0,0 +1,280 @@
+}
+
+func (a *AccessInfo) IsTokenExpired() bool {
+	if a.expiresAt.Before(time.Now()) {
+		return true
+	} else {
+		return false
+	}
+}
+
+func (a *AccessInfo) ShouldSkipAuthzCheckForNonAADUsers() bool {
+	return a.skipAuthzForNonAADUsers
+}
+
+func (a *AccessInfo) GetResultFromCache(request *authzv1.SubjectAccessReviewSpec, store authz.Store) (bool, bool) {
+	var result bool
+	key := getResultCacheKey(request)
+	glog.V(10).Infof("Cache search for key: %s", key)
+	found, _ := store.Get(key, &result)
+	return found, result
+}
+
+func (a *AccessInfo) SkipAuthzCheck(request *authzv1.SubjectAccessReviewSpec) bool {
+	if a.clusterType == connectedClusters {
+		_, ok := a.skipCheck[strings.ToLower(request.User)]
+		return ok
+	}
+	return false
+}
+
+func (a *AccessInfo) SetResultInCache(request *authzv1.SubjectAccessReviewSpec, result bool, store authz.Store) error {
+	key := getResultCacheKey(request)
+	glog.V(10).Infof("Cache set for key: %s, value: %t", key, result)
+	return store.Set(key, result)
+}
+
+func (a *AccessInfo) AllowNonResPathDiscoveryAccess(request *authzv1.SubjectAccessReviewSpec) bool {
+	if request.NonResourceAttributes != nil && a.allowNonResDiscoveryPathAccess && strings.EqualFold(request.NonResourceAttributes.Verb, "get") {
+		path := strings.ToLower(request.NonResourceAttributes.Path)
+		if strings.HasPrefix(path, "/api") || strings.HasPrefix(path, "/openapi") || strings.HasPrefix(path, "/version") || strings.HasPrefix(path, "/healthz") {
+			return true
+		}
+	}
+	return false
+}
+
+func (a *AccessInfo) CheckAccess(request *authzv1.SubjectAccessReviewSpec) (*authzv1.SubjectAccessReviewStatus, error) {
+	checkAccessBody, err := prepareCheckAccessRequestBody(request, a.clusterType, a.azureResourceId, a.retrieveGroupMemberships)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "error in preparing check access request")
+	}
+
+	checkAccessURL := *a.apiURL
+	// Append the path for azure cluster resource id
+	checkAccessURL.Path = path.Join(checkAccessURL.Path, a.azureResourceId)
+	exist, nameSpaceString := getNameSpaceScope(request)
+	if exist {
+		checkAccessURL.Path = path.Join(checkAccessURL.Path, nameSpaceString)
+	}
+
+	checkAccessURL.Path = path.Join(checkAccessURL.Path, checkAccessPath)
+	params := url.Values{}
+	params.Add("api-version", checkAccessAPIVersion)
+	checkAccessURL.RawQuery = params.Encode()
+
+	buf := new(bytes.Buffer)
+	if err := json.NewEncoder(buf).Encode(checkAccessBody); err != nil {
+		return nil, errors.Wrap(err, "error encoding check access request")
+	}
+
+	if glog.V(10) {
+		binaryData, _ := json.MarshalIndent(checkAccessBody, "", "    ")
+		glog.V(10).Infof("checkAccessURI:%s", checkAccessURL.String())
+		glog.V(10).Infof("binary data:%s", binaryData)
+	}
+
+	req, err := http.NewRequest(http.MethodPost, checkAccessURL.String(), buf)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating check access request")
+	}
+	// Set the auth headers for the request
+	req.Header = a.headers
```

Comment:
I think you need mutex.Lock() to read this map to avoid concurrent access to this map.